### ISSUE TITLE
Umple Generates Broken Code for Unique Immutable Attributes

### ISSUE THREAD

**Issue by jubiiz** [umple/Umple#2046]
## Description
When setting an attribute as 'unique immutable', the code generated by Umple fails to compile as it tries to use functions it has not defined. I have failed to find documentation about Umple explicitly not supporting setting an attribute as 'unique immutable'.
I'm aware that [the umple documentation](https://cruise.umple.org/umple/BeforeandAfterStatements.html) suggests using 'before' statements to prevent setting values (similar to immutable), but I'm inclined to think that the 'unique immutable' keyword pair should still work.

similar to : 
https://github.com/umple/umple/issues/400  (but in this issue the problem is that the code compiles and has functional bugs during runtime)

#### Setup: 
Bug encountered on Windows 10 Pro Education V.22H2
using Java openjdk 17.0.8.1 2023-08-24 LTS,
OpenJDK Runtime Environment Microsoft-8297089 (build 17.0.8.1+1-LTS),
OpenJDK 64-Bit Server VM Microsoft-8297089 (build 17.0.8.1+1-LTS, mixed mode, sharing),
coding on Microsoft Visual Studio Code V.1.72.2
with Umple plugin for Visual Studio Code, V1.2.1

Here is the directory structure of my setup: 
/unique_immutable/src/java/ contains: 
      UniqueImmutable.ump
      *(these two get automatically generated)*
      MyClass.java
      MyClass.class

#### Steps
1. Add the following code to `UniqueImmutable.ump`:
```
class MyClass{
    unique immutable myUniqueImmutableAttribute;
}
```
2. Generate the `.java` code with the `>Umple: Generate Code from Umple` command (ctrl+p to open the command palette). This step should succeed.
3. Attempt to compile the code with the `>Umple: Compile Code from Umple`.

## Expected results
a message showing success of compilation

## Actual Result
The following error was obtained: 
```
UniqueImmutable.ump:23: error: cannot find symbol
    if (!setMyUniqueImmutableAttribute(aMyUniqueImmutableAttribute))
         ^
  symbol:   method setMyUniqueImmutableAttribute(String)
  location: class MyClass
1 error
```
My understanding is that the 'immutable' keyword prevents a setter from being generated (`setMyUniqueImmutableAttribute()`) while the 'unique' keyword uses the setter to check for uniqueness, causing the problem.

**Comment by jubiiz** [comment]
This process works when I only put either of the 'unique' or 'immutable' keywords: the code generates and compiles as expected.

**Comment by jubiiz** [comment]
I have replicated this issue in Eclipse Version: 2023-06 (4.28.0), Build id: 20230608-1333
I was generating the code using the Umple plugin from http://cruise.eecs.uottawa.ca/org.cruise.umple.eclipse.plugin.update.site and compiling using 
```
javac MyClass.java
```
from the `/unique_immutable/src/java/` directory. 
Here again, I was able to make this work if I only put either of 'unique' or 'immutable'.

**Comment by TimLethbridge** [comment]
Thanks for this. I will assign it for fixing right away.

I confirm that if I paste your code into UmpleOnline, and hit 'Execute it', the same result appears.

Tijm

**Comment by TimLethbridge** [comment]
Design option 1: In the generator, detect the combined case of immutable+unique and put the relevant code from the set method in the constructor in ONLY that case. This may make the generation template a bit more complex. I think this option should be tried first. New tests will have to be generated, but this should not break many (if any) existing tests.

Design option 2: For ALL unique attributes, instead of calling the set method, put the relevant code from the set method in the constructor. Drawback: Would slightly bloat the code of cases where the attribute is unique but not immutable, which may be common. This change will break all existing tests for generation of unique; those tests would have to be changed, which involves some risk that they are changed incorrectly.


Option 3. Disallow unique immutable with an error. This would not be reasonable.

**Comment by TimLethbridge** [comment]
We discovered that if one uses 'unique lazy immutable' the code generated is much better ... the only issue is a missing argument to the constructor. So the suggested design approach (we can now call option 4) is this:

All 'lazy' does when combined with 'immutable' is that it gives one chance to set an immutable attribute. 

With three keywords together the lazy create a setter method, even when the attribute is immutable, but that setter can only be used once. And the constructor in the case of unique would make that call. ... except that the constructor does not currently have the aruemnt

1. In the parser, identify unique and immutable both being used together and set the relevant flag as though 'lazy' had been present, even if it was not.  This can be done at umpleInternalParser_CodeClass.ump analyzeAttribute line 4717

2. The relevant template is
  UmpleToJava/UmpleTLTemplates/constructor_AttributeAssignUnique.ump

3. We need to ensure the template generates the argument for the unique immutable attribute is adding an argument in the following code <<=gen.translate("parameterOne", av)>> Normally this will generate an argument, but it seems it is not doing so perhaps because of the presence of 'lazy' (which if used alone suppresses the presence of an argument on the constructor to allow it to be set later). We need to work out the logic to ensure the argument is not suppressed ... perhaps we need to add a new variable for this special sort of 'lazy' status.

4. The we need to add tests to make sure this works properly.

**Comment by TimLethbridge** [comment]
This was solved with PR #2055

**Comment by TimLethbridge** [comment]
Solved by PR #2055
